src/GraphQL/Validation/DocumentValidator.cs

```diff
@@ -134,5 +156,131 @@ private async Task<(IValidationResult validationResult, Variables variables)> Va
                 }
             }
         }
+
+        private class ParseArgumentVisitor : ASTVisitor<ParseArgumentVisitor.Context>
+        {
+            public static readonly ParseArgumentVisitor Instance = new();
+
+            private ParseArgumentVisitor() { }
+
+            protected override async ValueTask VisitOperationDefinitionAsync(GraphQLOperationDefinition operationDefinition, Context context)
+            {
+                var type = operationDefinition.Operation switch
+                {
+                    OperationType.Query => context.Schema.Query,
+                    OperationType.Mutation => context.Schema.Mutation,
+                    OperationType.Subscription => context.Schema.Subscription,
+                    _ => null
+                };
+                if (type == null)
+                    return;
+                context.Types.Push(type);
+                await base.VisitOperationDefinitionAsync(operationDefinition, context).ConfigureAwait(false);
+                context.Types.Pop();
+            }
+
+            protected override async ValueTask VisitInlineFragmentAsync(GraphQLInlineFragment inlineFragment, Context context)
+            {
+                var typeCondition = inlineFragment.TypeCondition;
+                if (typeCondition == null)
+                {
+                    await base.VisitInlineFragmentAsync(inlineFragment, context).ConfigureAwait(false);
+                }
+                else
+                {
+                    if (context.Schema.AllTypes[typeCondition.Type.Name.Value]?.GetNamedType() is IComplexGraphType type)
+                    {
+                        context.Types.Push(type);
+                        await base.VisitInlineFragmentAsync(inlineFragment, context).ConfigureAwait(false);
+                        context.Types.Pop();
+                    }
+                }
+            }
+
+            protected override async ValueTask VisitFragmentDefinitionAsync(GraphQLFragmentDefinition fragmentDefinition, Context context)
+            {
+                if (context.Schema.AllTypes[fragmentDefinition.TypeCondition.Type.Name.Value]?.GetNamedType() is IComplexGraphType type)
+                {
+                    context.Types.Push(type);
+                    await base.VisitFragmentDefinitionAsync(fragmentDefinition, context).ConfigureAwait(false);
+                    context.Types.Pop();
+                }
+            }
+
+            protected override async ValueTask VisitFieldAsync(GraphQLField field, Context context)
+            {
+                var fieldType = field.Name.Value == context.Schema.TypeMetaFieldType.Name
+                    ? context.Schema.TypeMetaFieldType
+                    : context.Type?.Fields.Find(field.Name.Value);
+                if (fieldType == null)
+                    return;
+                if (field.Arguments?.Count > 0)
```

Comment:
If there were no arguments defined on the field within the query, GetArguments will use the precalculated dictionary of default field argument values, so a dictionary does not need to be allocated.